add functionality to load Solcast API data to `iotools`

lorenzo-solcast opened the pull request:
- [x] Closes #1313 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

With this PR we are proposing to add a module `iotools.solcast` with functions to get irradiation and weather data for live, forecast, historic and tmy from the Solcast API. We noticed the demand for such functionality and many of our users use PVLib with Solcast data.

Note that Solcast has a free tier for radiation and weather data which you can access immediately by using a company email address when [signing up](https://toolkit.solcast.com.au/register). We also have unlimited requests at [unmetered locations](https://docs.solcast.com.au/#unmetered-locations) and additional free data is also [available](https://solcast.com/data-for-researchers) for researchers and students.

This is a first iteration and are looking for feedback, in particular about the following design decisions:

1. following the pattern in the other `iotools` modules, we have created a mapping to translate variables form Solcast's definition to PVLibs that is applied when `map_variables` is `True`. This is a 1-way translation so the user would still query the Solcast API with Solcast's definitions and get the variables in a `pd.DataFrame` with the variables converted to PVLib's. We could make the interface only using PVLib's definitions but that may cause confusion with the API documentation. 
2. testing is currently using `mocks`. If you prefer we can provide an API key to test directly against the API.
3. the function signatures have only the required parameters declared with all optional as `kwargs`. The intention is to avoid docstrings that are too verbose.

Some tests have been added but we'll add more after a first review. 

Thanks!

AdamRJensen commented:
@lorenzo-solcast thank you for this PR!

Note, we have not yet taken an official stance on iotools functions for commercial data providers so we can't promise this will get merged. Personally I'm in favor, under the condition that you take on the maintenance of the associated functions.

The irradiance iotools functions follow a pattern of returning a tuple of (data, meta). If there's no metadata to parse then just return an empty dictionary. Perhaps down the road metadata will be returned, and even if not then it's offered to have a standardized format.

Mock testing is probably sufficient (there's other iotools that only have that). Maybe it's even preferred.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,374 @@
+    map_variables: bool, default: True
+        When true, renames columns of the DataFrame to pvlib variable names
+        where applicable. See variable :const:`VARIABLE_MAP`.
+        Time is made the index with the "period mid" convention from Solcast's "period end".
```
Does this only happen when `map_variables` is True?

I think it would be preferable to always have the same time index convention, so I would move this comment to the output section.

lorenzo-solcast replied:
yes, it is applied only if `map_variables` is True. The idea being that the raw data is returned if set to `False`, but we can do that by default if preferred?

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,374 @@
+        where applicable. See variable :const:`VARIABLE_MAP`.
+        Time is made the index with the "period mid" convention from Solcast's "period end".
+    kwargs:
+        Optional parameters passed to the API. See https://docs.solcast.com.au/ for full list of parameters.
```
This doesn't render correctly in the documentation. I would suggest making a references section and then linking to that,  e.g.,

    Optional parameters passed to the API. See [1]_ for full list of parameters.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,374 @@
+
+    Returns
+    -------
+    df : pandas.DataFrame
+        containing the values for the parameters requested
```
```suggestion
    data : pandas.DataFrame
        containing the values for the parameters requested

```

Should also output a meta dictionary, even if empty

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,374 @@
+
+    Examples
+    --------
+    get_solcast_tmy(
```
This also does not render correctly in the documentation. Code needs to be indented four spaces.

Also, I would prefer if you added the `pvlib.iotools` to the functions, making it easier for new users to copy the code examples.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,374 @@
+    get_solcast_tmy(
+        latitude=-33.856784,
+        longitude=151.215297,
+        time_zone=10,
```
pythons time zone convention is a bit counterintuitive (see the [this link](https://pvlib-python.readthedocs.io/en/stable/user_guide/timetimezones.html)) so I would recommend adding a code comment stating what 10 corresponds to.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,374 @@
+    end : optional, datetime-like
+        Last day of the requested period
+    duration : optional, default is None
+        Must include one of end_date and duration. ISO_8601 compliant duration for the historic data.
```
Perhaps provide an example of what an ISO 8601 time duration is.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,374 @@
+        api_key="your-key"
+    )
+
+    use map_variables=False to avoid converting the data to PVLib's conventions
```
```suggestion
    use ``map_variables``=False to avoid converting the data to pvlib's conventions

```

AdamRJensen reviewed (commented):
An initial review from the airport😄

lorenzo-solcast commented:
thanks for the review @AdamRJensen! should have addressed most of your points and returning an empty `dict` to start with.

> Note, we have not yet taken an official stance on iotools functions for commercial data providers so we can't promise this will > get merged. Personally I'm in favor, under the condition that you take on the maintenance of the associated functions.

Well aware and totally up to the team. We'd be happy to contribute and take on the maintenance of the module.

lorenzo-solcast commented:
hi! any other thoughts on this PR?

cwhanse reviewed (approved):
Thanks @lorenzo-solcast 

@pvlib/pvlib-maintainer I support adding this.

lorenzo-solcast commented:
thanks for the review @cwhanse - your suggestion makes the docstring much clearer.

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+
+    Returns
+    -------
+    data : pandas.DataFrame
```
In all functions, the metadata dictionary should be listed in the `Returns` section as well.

lorenzo-solcast replied:
done.

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+        map_variables=map_variables
+    )
+
+    return data, {}
```
Is there no way to get metadata (data version, e.g.) back from the API?

Let's at least echo the `latitude` and `longitude` inputs in the metadata dict.

lorenzo-solcast replied:
data version not yet! I have added the lat/lon as suggested.

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+        where applicable. See variable :const:`VARIABLE_MAP`.
+    api_key : str
+        To access Solcast data you will need an API key [1]_.
+    kwargs:
+        Optional parameters passed to the GET request
+
+    See [2]_ for full list of parameters.
```
```suggestion
    kwargs:
        Optional parameters passed to the API.
        See [2]_ for full list of parameters.
```

For `kwargs`, we should be sure to be clear about the difference between GET parameters and parameters to `requests.get`.   Same for the other functions

lorenzo-solcast replied:
done.

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+
+    Examples
+    --------
+    >>> pvlib.iotools.solcast.get_solcast_tmy(
```
```suggestion
    >>> df, meta = pvlib.iotools.solcast.get_solcast_tmy(
```

And same for the other examples

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+    return data, {}
+
+
+def solcast2pvlib(data):
```
```suggestion
def _solcast2pvlib(data):
```

No need for this function to be public IMHO

lorenzo-solcast replied:
done.

kandersolar commented on pvlib/tests/iotools/test_solcast.py:
```diff
@@ -0,0 +1,122 @@
+import pandas as pd
+from pvlib.iotools.solcast import (
+    get_solcast_live, get_solcast_tmy, solcast2pvlib
```
We should test `_historical` and `_forecast` too, right?

lorenzo-solcast replied:
done.

AdamRJensen commented:
@lorenzo-solcast If you're able to address these issues within the week, we should be able to get these functions into the December pvlib release 🥳

lorenzo-solcast commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+    # azimuth -> solar_azimuth (degrees) (different convention)
+    ParameterMap(
+        "azimuth", "solar_azimuth", lambda x: abs(x) if x <= 0 else 360 - x
+    ),
```
done.

lorenzo-solcast commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+
+    Derived from satellite (clouds and irradiance over
+    non-polar continental areas) and numerical weather models (other data).
+    The TMY is calculated with data from 2007 to 2023.
```
added reference

lorenzo-solcast commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,443 @@
+        Must include either  ``end`` or ``duration``.
+        ISO_8601 compliant duration for the historic data,
+        like "P1D" for one day of data.
+        Must be within 31 days of the start_date.
```
done.

lorenzo-solcast commented:
thanks for the great review @kandersolar! should have addressed all the points you raised - let me know what you think.

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,448 @@
+    ParameterMap("precipitable_water", "precipitable_water", lambda x: x*10),
+    # zenith -> solar_zenith
+    ParameterMap("zenith", "solar_zenith")
+]
```
@AdamRJensen This is an interesting approach that we might consider adopting in some centralized way for iotools in general.

No action needed in this PR, of course.

AdamRJensen replied:
This is pretty cool

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,448 @@
+
+
+# define the conventions between Solcast and PVLib nomenclature and units
+VARIABLE_MAP = [
```
A few more renames are needed, I think:
- clearsky_dhi -> dhi_clear
- clearsky_dni -> dni_clear
- clearsky_ghi -> ghi_clear
- clearsky_gti -> poa_global_clear (??)

@AdamRJensen any thoughts on `clearsky_gti`?

AdamRJensen replied:
I suppose this is consistent with the usage of ``poa_global`` and ``_clear``. What are you thinking @kandersolar ?

kandersolar replied:
Was mostly wondering if we already had a name for clear-sky GTI, and if not, looking for a second opinion that `poa_global_clear` is a good choice :)

kandersolar commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,448 @@
+        where applicable. See variable :const:`VARIABLE_MAP`.
+    kwargs:
+        Optional parameters passed to the API.
+        See [3]_ for full list of parameters.
```
I think this should not be [3]

kandersolar reviewed (approved):
A couple more minor things but otherwise LGTM!

@AdamRJensen can you merge this when you think it's ready?

lorenzo-solcast commented:
@kandersolar thanks for spotting. Turns out that the Pandas version (0.25) used in that Python 3.7 workflow has an issue with parsing ISO8601 durations without days. I have implemented [a hacky fix](https://github.com/lorenzo-solcast/pvlib-python/blob/be27d850d4441d1a853ed1d7fab79233b9b467b8/pvlib/iotools/solcast.py#L387) that will work as we never provide monthly periods, but it is not ideal. 

Do you guys have an alternate/preferred/suggestion pattern to fix this?

kandersolar commented:
> Do you guys have an alternate/preferred/suggestion pattern to fix this?

I don't think we have a solution for this already (no other part of pvlib deals with ISO 8601 duration strings, to my knowledge).  That hacky fix seems reasonable to me.  If you want, you could leave a comment next to it indicating the better/simpler code we could switch to when we eventually increase our minimum pandas to 1.2.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+    you can pass any of the parameters listed in the API docs,
+    like asking for specific variables:
+
+    >>> df, meta = pvlib.iotools.solcast.get_solcast_forecast(
+    >>>     latitude=-33.856784,
+    >>>     longitude=151.215297,
+    >>>     output_parameters=['dni', 'clearsky_dni', 'snow_soiling_rooftop'],
+    >>>     api_key="your-key"
+    >>> )
+
```
I think it would be nice to add forecast horizon to this example using the ``hours`` parameter.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+    map_variables=True,
+    **kwargs
+):
+    """Get historical irradiance and weather estimated actuals
```
```suggestion
    """Get historical irradiance and weather estimated actuals.
```
I'm unfamiliar with the term "actuals" but maybe that is standard? As a non-native, I can't really tell what meaning is trying to be conveyed

lorenzo-solcast replied:
we refer to "live" and "historical" data as _estimated actuals_. These would be sometimes be called "measurements" I guess, but we avoid that term as we are really just estimating these values from satellite observations and not direct measurements of weather/power. Safe to assume that anyone using the Solcast API understands this?

AdamRJensen replied:
Hmm I think that's a very Solcast specific terminology. To be more in line with the other iotools I suggest removing mentions of "actuals" and just calling them "estimates"

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+        in decimal degrees, between -180 and 180, east is positive
+    start : datetime-like
+        First day of the requested period
+    end : optional, datetime-like
+        Last day of the requested period.
+        Must include one of ``end`` or ``duration``.
+    duration : optional, default is None
+        Must include either  ``end`` or ``duration``.
+        ISO_8601 compliant duration for the historic data,
+        like "P1D" for one day of data.
+        Must be within 31 days of the ``start``.
```
I think we went away from listing "default is None". In either case I would be consistent with the ``end`` and ``duration`` parameters which both are optional/defaults to None. I would prefer removing ", default is None"

AdamRJensen replied:
```suggestion
        Must be within 31 days of ``start``.
```

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+        Must include one of ``end`` or ``duration``.
+    duration : optional, default is None
+        Must include either  ``end`` or ``duration``.
+        ISO_8601 compliant duration for the historic data,
```
```suggestion
        ISO 8601 compliant duration for the historic data,
```

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+        parameters to be passed to the API
+    api_key : str
+        To access Solcast data you will need an API key [1]_.
+    map_variables: bool, default: True
+        When true, renames columns of the DataFrame to PVLib's variable names
+        where applicable. See variable :const:`VARIABLE_MAP`.
+        Time is the index as midpoint of each interval
+        from Solcast's "period end" convention.
```
Right now the documentation of the child functions ``map_variables`` parameter only states:
>  When true, renames columns of the DataFrame to pvlib variable names where applicable

I think that all instance of ``map_variables`` should also denote that the timestamp is shifted (I don't recall seeing this anywhere else in pvlib, but do support the idea).

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+def get_solcast_forecast(
+    latitude, longitude, api_key, map_variables=True, **kwargs
+):
+    """Get irradiance and weather forecasts from the present time
+    up to 14 days ahead
```
```suggestion
    """Get irradiance and weather forecasts from the present time
    up to 14 days ahead.
```

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+    latitude, longitude, api_key, map_variables=True, **kwargs
+):
+    """Get irradiance and weather estimated actuals for near real-time
+    and past 7 days
```
```suggestion
    and past 7 days.
```

AdamRJensen reviewed (approved):
Solid code! Only minor stuff from me.

Note, I wasn't able to test examples of the forecast/live data though with my API key.

lorenzo-solcast commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+
+    Examples
+    --------
+    >>> df, meta = pvlib.iotools.solcast.get_solcast_live(
+    >>>     latitude=-33.856784,
+    >>>     longitude=151.215297,
+    >>>    api_key="your-key"
+    >>> )
```
@AdamRJensen we have just given your account the required permissions to query for unmetered locations  if you want to test again!

lorenzo-solcast commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,459 @@
+    )
+
+    data = _get_solcast(
+        endpoint="tmy/radiation_and_weather",
```
that is a valid point, and relevant for our SDK too. There are no use-cases for this functionality yet, but I agree that down the line we may need expose this to the user.

AdamRJensen commented on pvlib/iotools/solcast.py:
```diff
@@ -0,0 +1,468 @@
+        api_key,
+        map_variables
+):
+    """retrieves weather, irradiance and power data from the Solcast API
```
```suggestion
    """Retrieve weather, irradiance and power data from the Solcast API.
```

AdamRJensen commented on pvlib/tests/iotools/test_solcast.py:
```diff
@@ -0,0 +1,215 @@
+import pandas as pd
+from pvlib.iotools.solcast import (
```
```suggestion
from pvlib.iotools import (
```
I would import it as a user would

AdamRJensen reviewed (approved):
I think we should be good to merge this function later today 🥳 

Great work @lorenzo-solcast!

AdamRJensen commented:
I would also recommend adding a "See Also" section to all the functions pointing to their Solcast cousins. Example:

    See Also
    --------
    pvlib.iotools.read_solaranywhere, pvlib.iotools.get_bsrn

AdamRJensen commented:
@lorenzo-solcast I made some minor changes so don't forget to pull from GH

I see that there are two lines that are not being hit by the tests, could you make some basic tests to hit these?
![image](https://github.com/pvlib/pvlib-python/assets/39184289/f4475966-0ddc-4bb5-ad0b-2ad39a62b4ac)

lorenzo-solcast commented:
@AdamRJensen added the tests!

AdamRJensen commented:
@lorenzo-solcast Access to Solcast data will now be supported by pvlib within a matter of days with the release of v0.10.3. Very exciting stuff! Thanks for your contribution and patience during the review process 🥳

lorenzo-solcast commented:
exciting indeed! looking forward to contribute further and thank you for the thorough reviews.
